File: 백준/Platinum/2494. 숫자 맞추기/숫자 맞추기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        int n = Integer.parseInt(br.readLine());
        int[][] dp = new int[n+1][10];
        int[][] seq = new int [n+1][10];
        int[] start = new int[n+1];
        int[] target = new int[n+1];
        String strStart = br.readLine();
        String strTarget = br.readLine();

        for (int i = 1; i <= n; i++) {
            start[i] = strStart.charAt(i - 1) - '0';
            target[i] = strTarget.charAt(i - 1) - '0';
        }

        for (int i = 1; i <= n; i++) {
            Arrays.fill(dp[i], 100000);
        }

        for (int i = 0; i < 10; i++) {
            dp[0][i] = i;
        }

        for (int i = 1; i < n+1; i++) {
            for (int j = 0; j < 10; j++) {
                int left = (target[i] - start[i] - j + 20) % 10;
                int right = 10 - left;
                int nextLeft = (j + left) % 10;

                // 오른쪽 회전
                if (dp[i][j] > dp[i-1][j] + right) {
                    dp[i][j] = dp[i-1][j] + right;
                    seq[i][j] = -right;
                }

                // 왼쪽 회전
                if (dp[i][nextLeft] > dp[i-1][j] + left) {
                    dp[i][nextLeft] = dp[i-1][j] + left;
                    seq[i][nextLeft] = left;
                }
            }
        }

        int result = 100000;
        int[] resultSeq = new int[n];
        int idx = 0;
        int cnt = n;

        for (int i = 0; i < 10; i++) {
            if (result > dp[n][i]) {
                result = dp[n][i];
                idx = i;
            }
        }

        sb.append(result + "\n"); // 총 회전횟수 출력

        while (cnt > 0) {
            int move = seq[cnt][idx];
            resultSeq[cnt-1] = move;

            if (move > 0) {
                idx = (idx - move + 10) % 10;
            }
            cnt --;
        }

        for (int i = 0; i < n; i++) {
            sb.append(i+1 + " " + resultSeq[i] + "\n");
        }
        System.out.println(sb);
    }
}

File: 백준/Platinum/2494. 숫자 맞추기/숫자 맞추기.py
# 숫자 맞추기
import sys
input = sys.stdin.readline

n = int(input())
start = [0] + [int(i) for i in input().rstrip()]
target = [0] + [int(i) for i in input().rstrip()]

dp = [[100000] * 10 for _ in range(n+1)]
seq = [[0] * 10 for _ in range(n+1)]

for i in range(10):
    dp[0][i] = i

for i in range(1, n+1):
    for j in range(10):
        left = (target[i] - start[i] - j + 20) % 10
        right = 10 - left
        nextLeft = (j + left) % 10
        
        # 오른쪽 회전
        if dp[i][j] > dp[i-1][j] + right:
            dp[i][j] = dp[i-1][j] + right
            seq[i][j] = -right # 회전횟수 갱신

        # 왼쪽 회전
        if dp[i][nextLeft] > dp[i-1][j] + left:
            dp[i][nextLeft] = dp[i-1][j] + left
            seq[i][nextLeft] = left # 회전횟수 갱신

result = 100000
resultSeq = []
cnt = n

for i in range(10):
    if result > dp[n][i]:
        result = dp[n][i]
        idx = i

print(result) # 총 회전횟수 출력

while cnt > 0:
    move = seq[cnt][idx]
    resultSeq.append(move)

    if move > 0:
        idx = (idx - move + 10) % 10

    cnt -= 1

for i in range(n):
    print(i+1, resultSeq[n-i-1]) # 숫자 나사번호별 회전횟수 출력
